Docker.DotNet/Microsoft.Net.Http.Client/BufferedReadStream.cs

```diff
@@ -0,0 +1,242 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Net.Http.Client
+{
+    internal class BufferedReadStream : Stream
+    {
+        private const char CR = '\r';
+        private const char LF = '\n';
+
+        private readonly Stream _inner;
+        private readonly byte[] _buffer;
+        private int _bufferOffset = 0;
+        private int _bufferCount = 0;
+        private bool _disposed;
+
+        public BufferedReadStream(Stream inner)
+        {
+            if (inner == null)
+            {
+                throw new ArgumentNullException("inner");
+            }
+            _inner = inner;
+            _buffer = new byte[1024];
+        }
+
+        public override bool CanRead
+        {
+            get { return _inner.CanRead || _bufferCount > 0; }
+        }
+
+        public override bool CanSeek
+        {
+            get { return false; }
+        }
+
+        public override bool CanTimeout
+        {
+            get { return _inner.CanTimeout; }
+        }
+
+        public override bool CanWrite
+        {
+            get { return _inner.CanWrite; }
+        }
+
+        public override long Length
+        {
+            get { throw new NotSupportedException(); }
+        }
+
+        public override long Position
+        {
+            get { throw new NotSupportedException(); }
+            set { throw new NotSupportedException(); }
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            _disposed = true;
+            if (disposing)
+            {
+                _inner.Dispose();
+            }
+        }
+
+        public override void Flush()
+        {
+            _inner.Flush();
+        }
+
+        public override Task FlushAsync(CancellationToken cancellationToken)
+        {
+            return _inner.FlushAsync(cancellationToken);
+        }
+
+        public override void Write(byte[] buffer, int offset, int count)
+        {
+            _inner.Write(buffer, offset, count);
+        }
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            return _inner.WriteAsync(buffer, offset, count, cancellationToken);
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            // Validate Inputs
+
+            // Drain buffer
+            if (_bufferCount > 0)
+            {
+                int toCopy = Math.Min(_bufferCount, count);
+                Buffer.BlockCopy(_buffer, _bufferOffset, buffer, offset, toCopy);
+                _bufferOffset += toCopy;
+                _bufferCount -= toCopy;
+                return toCopy;
+            }
```

Comment:
No, I think that would be incorrect. _inner.Read would block if there is currently no data available. We must return the available data immediately.